pkg/apis/pipeline/v1alpha1/build_gcs_resource.go

```diff
@@ -0,0 +1,155 @@
+// or a source manifest describing files to fetch.
+type BuildGCSResource struct {
+	Name           string
+	Type           PipelineResourceType
+	Location       string
+	DestinationDir string
+	ArtifactType   GCSArtifactType
+	//Secret holds a struct to indicate a field name and corresponding secret name to populate it
+	Secrets []SecretParam `json:"secrets"`
+}
+
+// NewBuildGCSResource creates a new BuildGCS resource to pass to knative build
+func NewBuildGCSResource(r *PipelineResource) (*BuildGCSResource, error) {
+	if r.Spec.Type != PipelineResourceTypeStorage {
+		return nil, fmt.Errorf("BuildGCSResource: Cannot create a BuildGCS resource from a %s Pipeline Resource", r.Spec.Type)
+	}
+	var location, destDir string
+	var aType GCSArtifactType
+	var locationSpecified, aTypeSpecified bool
+
+	for _, param := range r.Spec.Params {
+		switch {
+		case strings.EqualFold(param.Name, "Location"):
+			location = param.Value
+			if param.Value != "" {
+				locationSpecified = true
+			}
+		case strings.EqualFold(param.Name, "DestinationDir"):
+			destDir = param.Value
+		case strings.EqualFold(param.Name, "ArtifactType"):
+			aType = GCSArtifactType(param.Value)
+			if param.Value != "" {
+				aTypeSpecified = true
+			}
+		}
+	}
+
+	if !locationSpecified {
+		return nil, fmt.Errorf("BuildGCSResource: Need Location to be specified in order to create BuildGCS resource %s", r.Name)
+	}
+	if !aTypeSpecified {
+		return nil, fmt.Errorf("BuildGCSResource: Need ArtifactType to be specified in order to fetch BuildGCS resource %s", r.Name)
+	}
+	return &BuildGCSResource{
+		Name:           r.Name,
+		Type:           r.Spec.Type,
+		Location:       location,
+		DestinationDir: destDir,
+		ArtifactType:   aType,
+		Secrets:        r.Spec.SecretParams,
+	}, nil
+}
+
+// GetName returns the name of the resource
+func (s BuildGCSResource) GetName() string {
+	return s.Name
+}
+
+// GetType returns the type of the resource, in this case "storage"
+func (s BuildGCSResource) GetType() PipelineResourceType {
+	return PipelineResourceTypeStorage
+}
+
+// GetParams get params
+func (s *BuildGCSResource) GetParams() []Param { return []Param{} }
+
+// GetSecretParams returns the resource secret params
+func (s *BuildGCSResource) GetSecretParams() []SecretParam { return s.Secrets }
+
+// GetDownloadContainerSpec returns an array of container specs to download gcs storage object
+func (s *BuildGCSResource) GetDownloadContainerSpec() ([]corev1.Container, error) {
+	if s.DestinationDir == "" {
+		return nil, fmt.Errorf("BuildGCSResource: Expect Destination Directory param to be set %s", s.Name)
+	}
+	args := []string{"--type", string(s.ArtifactType), "--location", s.Location}
+	// dest_dir is the destination directory for GCS files to be copies"
+	if s.DestinationDir != "" {
+		args = append(args, "--dest_dir", filepath.Join(workspaceDir, s.DestinationDir))
+	}
+
+	envVars, secretVolumeMount := getSecretEnvVarsAndVolumeMounts(s.Name, gcsSecretVolumeMountPath, s.Secrets)
```

Comment:
im not sure we need the secret vars and volume mounts, do you know @shashwathi ? @tejal29 and I looked into it and didn't see any evidence that the fetcher image would make use of these mounted secrets

Reply:
So, after you left, i looked in to why the Secrets were not required in the existing [knative/build source code](https://github.com/knative/build/blob/master/pkg/reconciler/build/resources/pod.go) we were looking at.
In knative/build, MakePod would first the init-cred container would run which would make sure the credentials were set and the service accounts secrets are mounted. Then it would it would create the gcsToContainer.

In our case, since we first fetch the resource and then create a build for it, we need to fetch and mount the secrets. Does that make sense?

We can go over it again when we pair today.
Thanks
Tejal

Reply:
> In our case, since we first fetch the resource and then create a build for it, we need to fetch and mount the secrets. Does that make sense?

There was no support for GCS fetcher from private bucket in knative-build. Are we adding this feature in pipeline? @tejal29 @bobcatfish 

creds-init, git-init container were fetching docker and git secret configurations but not GCS. 
FYI I am okay with adding support as long as we document how a user needs to configure secrets in BuildGCS resource

Reply:
Our current GCS resource has [special support for GOOGLE_APPLICATION_CREDENTIALS](https://github.com/knative/build-pipeline/blob/master/docs/resources.md#storage-resource) - which I think makes sense, not sure if we need this for the fetcher image also?